### DISCUSSION TITLE
Practical approach for running cleanup code for child component before it's parent's cleanup.

### DISCUSSION THREAD

**Discussion by jsprog** [leptos-rs/leptos#1984] [Q&A]
I could confirm that [cleanup_fn](https://docs.rs/leptos/latest/leptos/fn.on_cleanup.html) of parent components are run before their children's counterparts. But sometimes the programmer need to reverse this order when using some third party libraries.

Question 1: How could we run cleanup code for children components before their parents' cleanup?
Question 2: How to hook into component's life cycle to run code just before removal of it's DOM content? 

Hint: Think about this as I'm trying to achieve onDestroy(() => { /* cleanup code */ }) from svelte.

**Comment by gbj** [comment]
It looks like the linked docs are wrong, for what it's worth, as they say that children are cleaned up first. I recall that switching this order fixed some kind of issue but I honestly can't remember what it was.

Question 2. Components don't have lifecycles, they are one-time setup functions. Cleanups follow the reactive graph, i.e., the nested dynamic parts of your app.

Question 1. There are simple ways to delay running some code (like wrapping `queue_microtask` around the parent's cleanup) but it's not clear to me whether they're going to be the right choice in your case. However, it may be that simple; again, depending on the use case.

The relevant function, in which parent cleanups run first. It's possible that running parent cleanups after children would actually be correct here, I would just hesitate to change it before I remember what the bug was that caused this ordering in the first place!

https://github.com/leptos-rs/leptos/blob/b1e8105442bf54948eb2c3290a21700c68885390/leptos_reactive/src/runtime.rs#L193-L212
